Story/110549 store document references

Rick-Ritense opened the pull request:
### Describe the changes
[Link to related target process story](https://ritense.tpondemand.com/entity/110549-be-temp-storage-metadata)

### Breaking changes
<!-- Valtimo aims to comply with the SemVer specification.  -->
<!-- Breaking changes are only allowed in the `next-major` branch.  -->
- [x] The contribution only contains changes that are not breaking.

### Documentation
<!-- Release notes should be available in the Valtimo documentation.  -->
- [x] Release notes have been written for these changes. 

Link to the pull request in the [Valtimo documentation repository](https://github.com/valtimo-platform/valtimo-documentation):
> https://github.com/valtimo-platform/valtimo-documentation/pull/711

New features or changes that have been introduced have been documented.
- [x] Yes
- [ ] Not applicable

### Tests
Unit tests have been added that cover these changes
- [x] Yes
- [ ] Not applicable

Integration tests have been added that cover these changes
- [x] Yes
- [ ] Not applicable

### Security
The [Secure by Default principle](https://en.wikipedia.org/wiki/Secure_by_default) has been applied to these changes
- [x] Yes
- [ ] Not applicable

Added or changed REST API endpoints have authentication and authorization in place
- [x] Yes
- [ ] Not applicable

Valtimo access control checks have been implemented
- [ ] Yes
- [x] Not applicable

### Dependencies
Newly added dependencies do not introduce known vulnerabilities/CVE's and are in line with the Valtimo license
- [ ] Yes
- [x] Not applicable

Commit:
feature 110549 - Initial commit

Create the JPA entity
Create the event and listener
The listener now stores the data in the database

Commit:
feature 110549 - Finished code

Created a new database table.
After saving the temporary document new execution variables are stored
These variables are stored in the database.
Created the ProcessBeans ResourceStorageDelegate and UploadProcessDelegate

Added the GetMapping to retrieve the metadata value

TODO: Fixing tests

Commit:
feature 110549 - Finished code

Using a more modern approach for the StorageMetadataKeys

Added multiple new tests

Commit:
Merge branch 'refs/heads/next-minor' into feature/110549-store-document-references

Rick-Ritense commented on resource/temporary-resource-storage/src/main/kotlin/com/ritense/resource/service/TemporaryResourceStorageService.kt:
```diff
@@ -19,28 +19,27 @@ package com.ritense.resource.service
     valtimoResourceTempDirectory: String = "",
     private val uploadProperties: ValtimoUploadProperties,
     private val objectMapper: ObjectMapper,
+    private val repository: ResourceStorageMetadataRepository
```
Does this change make it a breaking change or not?

Commit:
feature 110549 - Fixed failing tests

Commit:
feature 110549 - Fixed failing tests

Commit:
feature 110549 - Changed package name

I needed to add the @EnableJpaRepositories and @EntityScan on the TemporaryResourceStorageAutoConfiguration. But it turned out the same package names are also used in the deprecated module openzaak-resource (see: OpenZaakResourceAutoConfiguration)

After talking with Klaas about this we decided to change the package name for the classes that are scanned. Therefor the repository and domain packages now start with com.ritense.temporaryresource.

This is only done for the new classes so it can (hopefully) still ship as a non breaking change

Commit:
feature 110549 - Changed package name

Adjusted what fields are required for the ResourceStorageMetadataAvailableEvent

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/110549-store-document-references

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/110549-store-document-references

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/110549-store-document-references

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/110549-store-document-references

Klaas-Ritense commented on document-generation/smartdocuments/src/test/kotlin/com/ritense/smartdocuments/client/SmartDocumentsClientTest.kt:
```diff
@@ -23,52 +23,47 @@ import com.ritense.smartdocuments.domain.DocumentFormatOption
-import org.mockito.kotlin.verify
-import org.mockito.kotlin.whenever
+import org.mockito.Mockito.mock
+import org.mockito.kotlin.*
```
Could you change this and the other files back to single imports instead of the * imports?

Klaas-Ritense commented on resource/temporary-resource-storage/src/main/kotlin/com/ritense/resource/event/ResourceStorageMetadataAvailableEvent.kt:
This file and others are missing the copyright header

Klaas-Ritense commented on resource/temporary-resource-storage/src/main/kotlin/com/ritense/resource/service/TemporaryResourceStorageService.kt:
```diff
@@ -113,7 +110,30 @@ class TemporaryResourceStorageService(
+                    ).metadataValue
+                } catch (e: EntityNotFoundException) {
+                    logger.warn("Resource $resourceStorageFieldId does not exist: ${e.message}")
+                    throw e
```
This way, the error becomes part of the stacktrace:

```suggestion
                    throw IllegalStateException("Resource $resourceStorageFieldId does not exist", e)
```

Klaas-Ritense commented on resource/temporary-resource-storage/src/main/kotlin/com/ritense/resource/service/TemporaryResourceStorageService.kt:
```diff
@@ -113,7 +110,30 @@ class TemporaryResourceStorageService(
+            },
+            onFailure = { exception ->
+                logger.error("Failed to resolve metadata key '$metadataKey': ${exception.message}")
+                throw exception
```
```suggestion
                throw IllegalStateException("Failed to resolve metadata key '$metadataKey'", exception)
```

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/110549-store-document-references

Klaas-Ritense commented on zgw/documenten-api/src/main/kotlin/com/ritense/documentenapi/DocumentenApiPlugin.kt:
```diff
@@ -279,6 +274,27 @@ class DocumentenApiPlugin(
+            val pluginConfiguration = getDocumentenApiPluginByInformatieobjectUrl(URI.create(documentCreateResult.url))
+            execution.setVariable(DOWNLOAD_URL_PROCESS_VAR, createDownloadUrl(pluginConfiguration.id.id, documentId))
+        } catch (_: Exception) {
+            logger.warn { "Failed to set the $DOWNLOAD_URL_PROCESS_VAR variable in the DelegateExecution" }
```
```suggestion
            throw IllegalStateException("Failed to set the $DOWNLOAD_URL_PROCESS_VAR variable in the DelegateExecution")
```

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/110549-store-document-references

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/110549-store-document-references

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/110549-store-document-references

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/110549-store-document-references

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/110549-store-document-references

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/110549-store-document-references

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/110549-store-document-references

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/110549-store-document-references

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/110549-store-document-references

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/110549-store-document-references

Commit:
Applied code review findings

Improved error handling and added missing license headers.
Removed imports that were converted to .*

Commit:
Merge remote-tracking branch 'origin/story/110549-store-document-references' into story/110549-store-document-references

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/110549-store-document-references

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/110549-store-document-references

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/110549-store-document-references

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/110549-store-document-references

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/110549-store-document-references

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/110549-store-document-references

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/110549-store-document-references

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/110549-store-document-references

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/110549-store-document-references

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/110549-store-document-references

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/110549-store-document-references

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/110549-store-document-references

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/110549-store-document-references

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/110549-store-document-references

valtimo-platform[bot] committed:
Merge branch 'next-minor' into story/110549-store-document-references

Klaas-Ritense committed:
Merge branch 'next-minor' into story/110549-store-document-references

marijnritense committed:
Merge branch 'refs/heads/next-minor' into story/110549-store-document-references

# Conflicts:
#	resource/temporary-resource-storage/src/main/kotlin/com/ritense/resource/service/TemporaryResourceStorageService.kt

marijnritense committed:
imports

marijnritense committed:
Merge remote-tracking branch 'origin/story/110549-store-document-references' into story/110549-store-document-references

ivo-ritense commented on resource/temporary-resource-storage/build.gradle:
```diff
@@ -31,6 +32,8 @@ dependencies {
 
     testImplementation project(':test-utils-common')
 
+    api "org.camunda.bpm:camunda-engine:${camundaVersion}"
```
I think we generally want to restrict the camunda stuff to the core module. Do we actually need this?

marijnritense committed:
fixed test to work with the new code, fixed import, improved logging

marijnritense committed:
moved code around so zaak/document related classes are in the zaken API module. This also allows for the dependency removal of the camunda module for the temporarly resource storage module.
